Backend/editor/data/traversal.go

```diff
@@ -0,0 +1,84 @@
+package data
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+
+	"cms.csesoc.unsw.edu.au/editor/data/datamodels"
+)
+
+// Traverse returns the second last value and the last value pointed at by a specific path
+func Traverse(document datamodels.DataModel, subpaths []string) (reflect.Value, reflect.Value, error) {
+	prev := reflect.Value{}
+	curr := reflect.ValueOf(document)
+
+	for len(subpaths) > 0 {
+		prev, curr, subpaths = consumeField(prev, curr, subpaths)
+		prev, curr, subpaths = tryConsumeArrayElement(prev, curr, subpaths)
+
+		if !prev.IsValid() || !curr.IsValid() {
+			return reflect.Value{}, reflect.Value{}, fmt.Errorf("invalid path, couldn't find subpath %s", subpaths[0])
+		}
+	}
+
+	return prev, curr, nil
+}
+
+// tryConsumeArrayElement attempts to consume the head of the path, the assumption is that the current value is an
+// array value, if this is not the case then the function returns the prev, current and path uncahnged, if it was capable
+// of consuming the head then it returns the subpath minus the head
+func tryConsumeArrayElement(prev, curr reflect.Value, path []string) (reflect.Value, reflect.Value, []string) {
+	// If we are at an array, we attempt to "index" into that array directly, this implies consuming the next element of our path
+	// which should be a regular array index
+	atEndOfPath := len(path) < 1
+	canConsumeArrayIndex := !atEndOfPath &&
+		curr.IsValid() &&
+		(curr.Kind() == reflect.Array || curr.Kind() == reflect.Slice)
+
+	if canConsumeArrayIndex {
+		targetIndex := path[0]
+		field := getValueAtIndex(curr, targetIndex)
+		if !field.IsValid() {
+			return curr, reflect.Value{}, path
+		}
+
+		return curr, dereference(field), path[1:]
+	}
+
+	// Looks like we couldn't consume an array but thats ok :D
+	return prev, curr, path
+}
+
+// consumeField consumes a regular field from a path, it pops the head off the path slice and if it was able
+// to consume something then it retuns the path minus the head :D
+func consumeField(prev, curr reflect.Value, path []string) (reflect.Value, reflect.Value, []string) {
+	field := curr.FieldByName(path[0])
+	if curr.IsValid() {
+		return curr, dereference(field), path[1:]
+	}
+
+	return prev, reflect.Value{}, path
+}
+
+// getValueAtIndex fetches the value at the provided index in the array pointed at by reflect.Value
+// note it is assumed that index is a string
+func getValueAtIndex(array reflect.Value, i string) reflect.Value {
+	index, err := strconv.Atoi(i)
+	if err != nil || index >= array.Len() || index < 0 {
+		return reflect.Value{}
+	}
+
+	return array.Index(index)
```

Comment:
You may have to check if it's an array or a slice, since one requires `array.Elem().Index(index)` iirc. Maybe also change variable name from `array` to `list` or `l` or `xs` if you wanna be epic haskell man.

Reply:
No Haskell >:( , also the .Elem is if/when u allocate a array/slice with new, we're not really allocating anything so .Index works for both